sklearn/linear_model/base.py

```diff
@@ -264,6 +264,9 @@ def _set_intercept(self, X_offset, y_offset, X_scale):
             self.coef_ = self.coef_ / X_scale
             self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
         else:
+            if(len(self.coef_.shape) == 0):
+                self.coef_ = np.expand_dims(self.coef_, axis=0)
+            self.coef_ = np.array(self.coef_)
```

Comment:
sorry I don't get the fix.

also:
```
if(len(self.coef_.shape) == 0):
```
should be written
```
if self.coef_.ndim == 0:
```

if it's a bad use of squeeze above it's there that it should be fixed. not here.

Reply:
Ok, I will try to resolve the issue

Reply:
Hi @agramfort and @CoderINusE , I'm a new comer and I just checked it out.

I used this test:

```
def test_elastic_net_no_intercept_coef_shape():
    X = [[-1], [0], [1]]
    y = [-1, 0, 1]

    for intercept in [True, False]:
        clf = ElasticNet(fit_intercept=intercept)
        clf.fit(X, y)
        coef_ = clf.coef_
        assert_equal(coef_.shape, (1,))
```

the lines I debugged in ElasticNet.fit() 

``` 
        import pdb; pdb.set_trace()
        self.coef_, self.dual_gap_ = map(np.squeeze, [coef_, dual_gaps_])

        self._set_intercept(X_offset, y_offset, X_scale)

        # workaround since _set_intercept will cast self.coef_ into X.dtype
        self.coef_ = np.asarray(self.coef_, dtype=X.dtype)
        
        # return self for chaining fit and predict calls
        return self
```

and here's the results of debugging: 

```
test_coordinate_descent.py 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>


******** with intercept

> /scikit-learn/sklearn/linear_model/coordinate_descent.py(763)fit()
-> self.coef_, self.dual_gap_ = map(np.squeeze, [coef_, dual_gaps_])
(Pdb) coef_
array([[ 0.14285714]]) # before np.squeeze

(Pdb) next
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(765)fit()
-> self._set_intercept(X_offset, y_offset, X_scale)
(Pdb) self.coef_
array(0.14285714285714285) # after np.squeeze

(Pdb) next
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(768)fit()
-> self.coef_ = np.asarray(self.coef_, dtype=X.dtype)
(Pdb) self.coef_
array([ 0.14285714]) # after set_intercept

(Pdb) next
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(771)fit()
-> return self
(Pdb) self.coef_
array([ 0.14285714]) # after np.asarray


******** without intercept

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(763)fit()
-> self.coef_, self.dual_gap_ = map(np.squeeze, [coef_, dual_gaps_])
(Pdb) coef_
array([[ 0.14285714]]) # before np.squeeze

(Pdb) next
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(765)fit()
-> self._set_intercept(X_offset, y_offset, X_scale)
(Pdb) self.coef_
array(0.14285714285714285) # after np.squeeze

(Pdb) next
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(768)fit()
-> self.coef_ = np.asarray(self.coef_, dtype=X.dtype)
(Pdb) self.coef_
array(0.14285714285714285) after set_intercept

(Pdb) next
> /scikit-learn/sklearn/linear_model/coordinate_descent.py(771)fit()
-> return self
(Pdb) self.coef_
array(0.14285714285714285) # after np.asarray
```
so if the test case I used is correct it seems like what causes this (or doesn't handle the case) is `base.LinearModel._set_intercept`
```
    def _set_intercept(self, X_offset, y_offset, X_scale):
        """Set the intercept_
        """
        if self.fit_intercept:
            self.coef_ = self.coef_ / X_scale
            self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
        else:
            self.intercept_ = 0.
```
I think it's related to the broadcasting occurring at `self.coef_ = self.coef_ / X_scale ` , which doesn't happen in the second case.

Reply:
Hi @CoderINusE, please take a look on my analysis at [#10571](https://github.com/scikit-learn/scikit-learn/issues/10571), according to my analysis this bug isn't caused by `np.squeeze`. The call to `self._set_intercept(X_offset, y_offset, X_scale)` in `ElasticNet.fit(..)` should handle the shape of `self.coef_` but it doesn't (in the case of `fit_intercept=False`).

There are two options to resolve this as I see it:

1. Modify `base._set_intercept(..)` to handle the shape. However changing this function might cause side effects then you should make sure it doesn't fail tests of other modules (you can find the corresponding tests, or just run all of them).

2. Go back to `master` version of `base._set_intercept`, then bypass this bug through `ElasticNet.fit()` . Right after the call to  `_set_intercept(..)`. This way you'll only have to test `coordinate_descent.py`.

Hope this helps!